### PR TITLE
fix(dsraft): avoid allocating shards on lost sites

### PR DESCRIPTION
**Pull request by keynslug**
Part of [EMQX-13864](https://emqx.atlassian.net/browse/EMQX-13864).

Release version: 5.9

## Summary

Address an edge case in shard allocation: avoid allocating shards to lost sites. This is extremely unlikely to happen in practice, but occasionally manifests in emqx/emqx-operator#1123 test runs.

See individual commits for details.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] ~~Added property-based tests for code which performs user input validation~~
- [ ] Changed lines covered by tests
- [x] Change log has been added to `changes/ee/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] ~~Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket~~
- [x] Schema changes are backward compatible

[EMQX-13864]: https://emqx.atlassian.net/browse/EMQX-13864?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx_ds_builtin_raft/src/emqx_ds_replication_layer_meta.erl [hunk 1]
```diff
@@ -956,6 +966,19 @@ forget_node(Node) ->
             logger:error("Failed to forget leaving node ~p: ~p", [Node, Reason])
     end.
 
+forget_lost_sites() ->
+    NodesLost = [R || R = #?NODE_TAB{node = N} <- all_nodes(), node_status(N) =:= lost],
```
Comment (thalesmg) [inline apps/emqx_ds_builtin_raft/src/emqx_ds_replication_layer_meta.erl:970]: Tiny nit:

```suggestion
    NodesLost = sites(lost),
```

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_ds_builtin_raft/src/emqx_ds_replication_layer_meta.erl (hunk 1 above)
Comment (keynslug) [reply apps/emqx_ds_builtin_raft/src/emqx_ds_replication_layer_meta.erl:970]: Yeah but the thing is we need whole records later, not just `site()`s. 🫠

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-6840)

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-6840)
